src/normalization.md

```diff
@@ -0,0 +1,314 @@
+
+Finally, an alias can just be unnormalized, `<Vec<u32> as IntoIterator>::Iter` is an unnormalized alias as it can already be normalized to `std::vec::IntoIter<u32>`, it just hasn't been done yet.
+
+---
+
+It is worth noting that Free and Inherent aliases cannot be rigid or ambiguous as   naming them also implies having resolved the definition of the alias, which specifies the underlying type of the alias:
+```rust
+#![feature(inherent_associated_types)]
+
+trait Trait {}
+
+impl Trait for u32 {}
+impl Trait for u64 {}
+
+struct Foo<T>(T);
+
+impl<T: Trait> Foo<T> {
+    type Inherent = u32;
+}
+
+fn ret_alias<T: Trait>(_: T) -> Foo<T>::Inherent {
+    10_u32
+}
+
+fn main() {
+    // `t` has type `?x`
+    let mut t = Default::default();
+    // `ret_alias` returns a value of type `Foo<?x>::Inherent`
+    t = ret_alias(t);
+    
+    // we have equated `?x` with `Foo<?x>::Inherent` this can
+    // only succeed if we're able to normalize inherent aliases
+    // with ambiguous where clauses on the impl.
+    //
+    // Or, put another way, it is *not* an ambiguous alias.
+}
+```
+
+### Diverging Aliases
+
+An alias is considered to "diverge" if its definition does not actually specify an underlying non-alias type to normalize to. On the flip side, we consider an alias to "terminate" if it does have an underlying value it can be normalized to (i.e. it does not diverge). A concrete example of diverging aliases:
+```rust
+type Diverges = Diverges;
+
+trait Trait {
+    type DivergingAssoc;
+}
+impl Trait for () {
+    type DivergingAssoc = <() as Trait>::DivergingAssoc;
+}
+```
+In this example both `Diverges` and `DivergingAssoc` are "trivial" cases of diverging type aliases where they have been defined as being equal to themselves. There is no underlying type that `Diverges` can ever be normalized to.
+
+We generally try to error when diverging aliases are defined, but this is entirely a "best effort" check. In the previous example the definitions are "simple enough" to be detected and so errors are emitted. However, in more complex cases, or cases where only some instantiations of generic parameters would result in a diverging alias, we don't emit an error:
+```rust
+trait Trait {
+    type DivergingAssoc<U: Trait>;
+}
+impl<T: ?Sized> Trait for T {
+    // This alias always diverges but we don't emit an error because
+    // the compiler can't "see" that.
+    type DivergingAssoc<U: Trait> = <U as Trait>::DivergingAssoc<U>;
+}
+```
+
+Ultimately this means that we have no guarantee that aliases in the type system terminate. As aliases may only diverge for some specific arguments to generic parameters, it also means that only fully concrete aliases can be checked to terminate, which means that codegen/const evaluation is the only time that aliases can reliably be checked to terminate:
```

Comment:
```suggestion
Ultimately this means that we have no guarantee that aliases in the type system don't diverge. As aliases may only diverge for some specific arguments to generic parameters, it also means that we only know whether an alias diverges once it is fully concrete. This means that codegen/const evaluation also has to handle diverging aliases can reliably be checked to terminate:
```
I personally dislike the perspective "only codegen can check". It's more that "codegen has to check for them *as well*"

Reply:
I'm not really sure I understand the distinction :thinking: Codegen has to check for them because *only* codegen can check some aliases.

Reply:
I updated the wording anyway :thinking: